src/uct/ib/base/ib_device.c

```diff
@@ -527,21 +529,29 @@ ucs_status_t uct_ib_device_query(uct_ib_device_t *dev,
         break;
     }
 
-    if (dev->num_ports > UCT_IB_DEV_MAX_PORTS) {
-        ucs_error("%s has %d ports, but only up to %d are supported",
-                  ibv_get_device_name(ibv_device), dev->num_ports,
-                  UCT_IB_DEV_MAX_PORTS);
-        return UCS_ERR_UNSUPPORTED;
-    }
-
+    num_ports = dev->num_ports < UCT_IB_DEV_MAX_PORTS ? dev->num_ports : UCT_IB_DEV_MAX_PORTS;
     /* Query all ports */
-    for (i = 0; i < dev->num_ports; ++i) {
+    for (i = 0; i < num_ports; ++i) {
         ret = ibv_query_port(dev->ibv_context, i + dev->first_port,
                              &dev->port_attr[i]);
         if (ret != 0) {
             ucs_error("ibv_query_port() returned %d: %m", ret);
             return UCS_ERR_IO_ERROR;
         }
+        if (dev->port_attr[i].gid_tbl_len > 0) {
+            has_gid = 1;
+            break;
+        }
+    }
+    if (has_gid == 0) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (dev->num_ports > UCT_IB_DEV_MAX_PORTS) {
+        ucs_error("%s has %d ports, but only up to %d are supported",
+                  ibv_get_device_name(ibv_device), dev->num_ports,
+                  UCT_IB_DEV_MAX_PORTS);
+        return UCS_ERR_UNSUPPORTED;
```

Comment:
let's move this to line 532: if num ports is > MAX_PORTS, print warning, but continue using only 2 ports, and NOT fail.

Reply:
@yosefe So for BF2/ARM switchdev mode, it will always have warnings. Is this good?

Reply:
so let's make it ucs_debug (not ucs_warn)